File: src/server/api/profiles.ts
/**
 * Profiles API
 * GET /api/profiles - List all profiles
 * GET /api/profiles/:name - Get single profile
 * POST /api/profiles - Create profile
 * PUT /api/profiles/:name - Update profile
 * DELETE /api/profiles/:name - Delete profile
 */

import type { IncomingMessage, ServerResponse } from "node:http";
import type { ClaudeCodeProfile } from "../../types.js";
import * as manager from "../../config/manager.js";
import { sendError, sendJson } from "../middleware.js";
import { sanitizeProfile, fullProfile } from "./util.js";
import { getAllPresets } from "../../providers/presets.js";

/**
 * GET /api/profiles - List all profiles
 */
export async function listProfiles(req: IncomingMessage, res: ServerResponse): Promise<void> {
  const profiles = await manager.listProfiles();
  sendJson(res, profiles.map(sanitizeProfile));
}

/**
 * GET /api/profiles/:name - Get single profile
 */
export async function getProfile(req: IncomingMessage, res: ServerResponse, params: Record<string, string>): Promise<void> {
  const { name } = params;

  if (!name) {
    sendError(res, { code: "INVALID_PARAMS", message: "Profile name is required" }, 400);
    return;
  }

  const profile = await manager.getProfile(name);

  if (!profile) {
    sendError(res, { code: "PROFILE_NOT_FOUND", message: `Profile "${name}" does not exist` }, 404);
    return;
  }

  sendJson(res, sanitizeProfile(profile));
}

/**
 * POST /api/profiles - Create profile
 */
export async function createProfile(req: IncomingMessage, res: ServerResponse): Promise<void> {
  const body = (req as any).body;

  if (!body || !body.name || !body.providerId) {
    sendError(res, { code: "INVALID_PARAMS", message: "name and providerId are required" }, 400);
    return;
  }

  // Verify provider exists
  const providers = await getAllPresets();
  const provider = providers.find(p => p.id === body.providerId);

  if (!provider) {
    sendError(res, { code: "UNKNOWN_PROVIDER", message: `Provider "${body.providerId}" not found` }, 400);
    return;
  }

  // Check if profile already exists
  const existing = await manager.getProfile(body.name);
  if (existing) {
    sendError(res, { code: "PROFILE_EXISTS", message: `Profile "${body.name}" already exists` }, 409);
    return;
  }

  try {
    const now = new Date().toISOString();
    const profile: ClaudeCodeProfile = {
      name: body.name,
      providerId: body.providerId,
      apiKey: body.apiKey || "",
      authToken: body.authToken || "",
      baseURL: body.baseURL || provider.baseURL,
      model: body.model,
      openaiModel: body.openaiModel,
      models: body.models,
      envKey: body.envKey,
      headers: body.headers,
      createdAt: now,
      updatedAt: now,
    };

    await manager.upsertProfile(profile, body.coder);

    sendJson(res, sanitizeProfile(profile), 201);
  } catch (error) {
    sendError(res, { code: "CREATE_FAILED", message: error instanceof Error ? error.message : "Failed to create profile" }, 500);
  }
}

/**
 * PUT /api/profiles/:name - Update profile
 */
export async function updateProfile(req: IncomingMessage, res: ServerResponse, params: Record<string, string>): Promise<void> {
  const { name } = params;
  const body = (req as any).body;

  if (!name) {
    sendError(res, { code: "INVALID_PARAMS", message: "Profile name is required" }, 400);
    return;
  }

  // Check if profile exists
  const existing = await manager.getProfile(name);
  if (!existing) {
    sendError(res, { code: "PROFILE_NOT_FOUND", message: `Profile "${name}" does not exist` }, 404);
    return;
  }

  try {
    const now = new Date().toISOString();
    const profile: ClaudeCodeProfile = {
      ...existing,
      ...body,
      name, // Ensure name stays consistent with URL param
      updatedAt: now,
    };

    await manager.upsertProfile(profile);

    sendJson(res, sanitizeProfile(profile));
  } catch (error) {
    sendError(res, { code: "UPDATE_FAILED", message: error instanceof Error ? error.message : "Failed to update profile" }, 500);
  }
}

/**
 * DELETE /api/profiles/:name - Delete profile
 */
export async function deleteProfile(req: IncomingMessage, res: ServerResponse, params: Record<string, string>): Promise<void> {
  const { name } = params;

  if (!name) {
    sendError(res, { code: "INVALID_PARAMS", message: "Profile name is required" }, 400);
    return;
  }

  try {
    await manager.deleteProfile(name);
    sendJson(res, { success: true, message: `Profile "${name}" deleted` });
  } catch (error) {
    sendError(res, { code: "DELETE_FAILED", message: error instanceof Error ? error.message : "Failed to delete profile" }, 500);
  }
}

File: packages/cli/crates/server/src/server/routes/profiles.rs
//! TS: server/api/profiles.ts
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
    routing::get,
    Json, Router,
};
use swixter_core::types::{now_iso, Profile, ProviderPreset};

use crate::server::error::ApiError;
use crate::server::state::AppState;
use crate::server::util::sanitize_profile;

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/profiles", get(list_profiles).post(create_profile))
        .route(
            "/profiles/{name}",
            get(get_profile).put(update_profile).delete(delete_profile),
        )
}

/// provider 查找：用户自定义（state 注入路径）优先，其次内置 preset（TS getProviderById 语义）
pub fn find_provider(state: &AppState, id: &str) -> Option<ProviderPreset> {
    if let Some(p) = swixter_core::user_providers::load_from(&state.providers_path())
        .into_iter()
        .find(|p| p.id == id)
    {
        return Some(p);
    }
    swixter_core::presets::find_builtin(id).cloned()
}

async fn list_profiles(State(state): State<AppState>) -> impl IntoResponse {
    let mgr = state.config_manager();
    let list: Vec<Profile> = mgr
        .config()
        .profiles
        .values()
        .map(sanitize_profile)
        .collect();
    Json(list)
}

async fn get_profile(
    State(state): State<AppState>,
    Path(name): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    let mgr = state.config_manager();
    let profile = mgr.get_profile(&name).ok_or_else(|| {
        ApiError::not_found(
            "PROFILE_NOT_FOUND",
            format!("Profile \"{name}\" does not exist"),
        )
    })?;
    Ok(Json(sanitize_profile(profile)))
}

async fn create_profile(
    State(state): State<AppState>,
    Json(body): Json<serde_json::Value>,
) -> Result<impl IntoResponse, ApiError> {
    let name = body.get("name").and_then(|v| v.as_str());
    let provider_id = body.get("providerId").and_then(|v| v.as_str());
    let (Some(name), Some(provider_id)) = (name, provider_id) else {
        return Err(ApiError::bad_request(
            "INVALID_PARAMS",
            "name and providerId are required",
        ));
    };

    let provider = find_provider(&state, provider_id).ok_or_else(|| {
        ApiError::bad_request(
            "UNKNOWN_PROVIDER",
            format!("Provider \"{provider_id}\" not found"),
        )
    })?;

    let mut mgr = state.config_manager();
    if mgr.get_profile(name).is_some() {
        return Err(ApiError::conflict(
            "PROFILE_EXISTS",
            format!("Profile \"{name}\" already exists"),
        ));
    }

    // TS createProfile：未传字段回退默认值；baseURL 缺省继承 provider
    let now = now_iso();
    let mut v = serde_json::json!({
        "name": name,
        "providerId": provider_id,
        "apiKey": body.get("apiKey").and_then(|x| x.as_str()).unwrap_or(""),
        "baseURL": body
            .get("baseURL")
            .and_then(|x| x.as_str())
            .unwrap_or(&provider.base_url),
        "createdAt": now,
        "updatedAt": now,
    });
    if let Some(t) = body
        .get("authToken")
        .and_then(|x| x.as_str())
        .filter(|s| !s.is_empty())
    {
        v["authToken"] = t.into();
    }
    for k in ["model", "openaiModel", "models", "envKey", "headers"] {
        if let Some(x) = body.get(k) {
            v[k] = x.clone();
        }
    }
    let profile: Profile = serde_json::from_value(v)
        .map_err(|e| ApiError::bad_request("INVALID_PARAMS", e.to_string()))?;

    let coder = body.get("coder").and_then(|x| x.as_str());
    mgr.upsert_profile(profile.clone(), coder)
        .map_err(|e| ApiError::internal("CREATE_FAILED", e.to_string()))?;

    Ok((StatusCode::CREATED, Json(sanitize_profile(&profile))))
}

async fn update_profile(
    State(state): State<AppState>,
    Path(name): Path<String>,
    Json(body): Json<serde_json::Value>,
) -> Result<impl IntoResponse, ApiError> {
    let mut mgr = state.config_manager();
    let existing = mgr.get_profile(&name).cloned().ok_or_else(|| {
        ApiError::not_found(
            "PROFILE_NOT_FOUND",
            format!("Profile \"{name}\" does not exist"),
        )
    })?;

    // TS: {...existing, ...body, name, updatedAt: now} —— name 取 URL 参数
    let mut v = serde_json::to_value(&existing)
        .map_err(|e| ApiError::internal("UPDATE_FAILED", e.to_string()))?;
    if let Some(obj) = body.as_object() {
        for (k, val) in obj {
            v[k] = val.clone();
        }
    }
    v["name"] = name.clone().into();
    v["updatedAt"] = now_iso().into();
    let profile: Profile = serde_json::from_value(v)
        .map_err(|e| ApiError::bad_request("INVALID_PARAMS", e.to_string()))?;

    mgr.upsert_profile(profile.clone(), None)
        .map_err(|e| ApiError::internal("UPDATE_FAILED", e.to_string()))?;
    Ok(Json(sanitize_profile(&profile)))
}

async fn delete_profile(
    State(state): State<AppState>,
    Path(name): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    let mut mgr = state.config_manager();
    mgr.delete_profile(&name)
        .map_err(|e| ApiError::internal("DELETE_FAILED", e.to_string()))?;
    Ok(Json(serde_json::json!({
        "success": true,
        "message": format!("Profile \"{name}\" deleted"),
    })))
}
